src/main/kotlin/com/terraformation/backend/documentproducer/VariableService.kt

```diff
@@ -0,0 +1,44 @@
+package com.terraformation.backend.documentproducer
+
+import com.terraformation.backend.customer.model.requirePermissions
+import com.terraformation.backend.db.docprod.DocumentId
+import com.terraformation.backend.documentproducer.db.DocumentStore
+import com.terraformation.backend.documentproducer.db.VariableStore
+import com.terraformation.backend.documentproducer.db.VariableValueStore
+import com.terraformation.backend.documentproducer.model.SectionValueVariable
+import com.terraformation.backend.documentproducer.model.SectionVariable
+import com.terraformation.backend.documentproducer.model.Variable
+import jakarta.inject.Named
+
+@Named
+class VariableService(
+    private val documentStore: DocumentStore,
+    private val variableStore: VariableStore,
+    private val variableValueStore: VariableValueStore,
+) {
+  fun fetchDocumentVariables(documentId: DocumentId): List<Variable> {
+    requirePermissions { readDocument(documentId) }
+
+    val document = documentStore.fetchDocumentById(documentId)
+    val manifestVariables = variableStore.fetchManifestVariables(document.variableManifestId!!)
+
+    val sectionVariableIds =
+        manifestVariables.flatMap { parent ->
+          when (parent) {
+            is SectionVariable -> parent.children.map { child -> child.id }
+            else -> emptyList()
+          }
+        }
+
+    val injectedSectionValues =
+        variableValueStore
+            .listValues(projectId = document.projectId!!, variableIds = sectionVariableIds)
+            .map { it.value }
+            .filterIsInstance<SectionValueVariable>()
+
+    val injectedSectionValueVariables =
+        injectedSectionValues.map { variableStore.fetchVariable(it.usedVariableId) }
```

Comment:
I wonder if it makes sense to do a custom fetch method to get the referenced variables. This process is going to fetch a bunch of data it doesn't end up needing, and do more round trips to the database than may be needed.

Maybe something like
```kotlin
  fun fetchUsedVariableIds(documentId: DocumentId): Set<VariableId> {
    return dslContext
        .select(
            VARIABLE_VALUES.VARIABLE_ID,
            VARIABLE_VALUES.LIST_POSITION,
            VARIABLE_SECTION_VALUES.USED_VARIABLE_ID)
        .distinctOn(
            VARIABLE_VALUES.VARIABLE_ID,
            VARIABLE_VALUES.LIST_POSITION)
        .from(VARIABLE_VALUES)
        .join(VARIABLE_SECTION_VALUES)
        .on(VARIABLE_VALUES.ID.eq(VARIABLE_SECTION_VALUES.VARIABLE_VALUE_ID))
        .where(VARIABLE_SECTION_VALUES.USED_VARIABLE_ID.isNotNull)
        .orderBy(
            VARIABLE_VALUES.VARIABLE_ID,
            VARIABLE_VALUES.LIST_POSITION,
            VARIABLE_VALUES.ID.desc())
        .fetchSet(VARIABLE_SECTION_VALUES.USED_VARIABLE_ID.asNonNullable())
  }
```
to fetch the referenced IDs, and then a new `VariableStore.fetchVariables()` that takes a collection of variable IDs and fetches them all at once?

Reply:
I will give that a go, thanks.

Reply:
I had to change it up a bit to include the document ID in my latest commit.